src/Toml/Codec/Combinators.hs

```diff
@@ -516,34 +520,112 @@ map keyCodec valCodec key = Codec input output
+if your 'Map' keys are types like 'Text' and you want to work with raw
+TOML keys directly.
+
+For example, if you have TOML like this:
+
+@
+[colours]
+yellow = "#FFFF00"
+red    = { red = 255, green = 0, blue = 0 }
+pink   = "#FFC0CB"
+@
+
+You want to convert such TOML configuration into the following Haskell
+types:
+
+
+@
+__data__ Rgb = Rgb
+    { rgbRed   :: Int
+    , rgbGreen :: Int
+    , rgbBlue  :: Int
+    }
+
+__data__ Colour
+   = Hex Text
+   | RGB Rgb
+
+colourCodec :: 'TomlCodec' Colour
+colourCodec = ...
+
+__data__ ColourConfig = ColourConfig
+    { configColours :: 'Map' 'Text' Colour
+    }
+@
+
+And you want in the result to have a 'Map' like this:
+
+@
+'Map.fromList'
+    [ "yellow" -> Hex "#FFFF00"
+    , "ping"   -> Hex "#FFC0CB"
+    , "red"    -> Rgb 255 0 0
+    ]
+@
+
+You can use 'tableMap' to define 'TomlCodec' in the following way:
+
+@
+colourConfigCodec :: 'TomlCodec' ColourConfig
+colourConfigCodec = ColourConfig
+    \<$\> Toml.'tableMap' Toml._KeyText colourCodec "colours" .= configColours
+@
+
+__Hint:__ You can use 'Toml.Codec.BiMap._KeyText' or
+'Toml.Codec.BiMap._KeyString' to convert betwen TOML keys and 'Map'
+keys (or you can write your custom 'TomlBiMap').
+
+__NOTE__: Unlike the 'map' codec, this codec is less flexible (i.e. it doesn't
+allow to have arbitrary structures as 'Key's, it works only for
+text-like keys), but can be helpful if you want to save a few
+keystrokes during TOML configuration. A similar TOML configuration,
+but suitable for the 'map' codec will look like this:
+
+@
+colours =
+    [ { key = "yellow", hex = "#FFFF00" }
+    , { key = "pink",   hex = "#FFC0CB" }
+    , { key = "red",    rgb = { red = 255, green = 0, blue = 0 } }
+    ]
+@
+
+@since 1.3.0.0
+-}
+tableMap
+    :: forall k v
+    .  Ord k
+    => TomlBiMap Key k
+    -- ^ Bidirectional converter between TOML 'Key's and 'Map' keys
+    -> (Key -> TomlCodec v)
+    -- ^ Codec for 'Map' values for the corresponding 'Key'
+    -> Key
+    -- ^ Table name for 'Map'
+    -> TomlCodec (Map k v)
+tableMap keyBiMap valCodec tableName = Codec input output
   where
-    input :: TomlEnv (Map key val)
-    input = do
-        mTable <- asks $ Prefix.lookup key . tomlTables
-        case mTable of
-            Nothing   -> pure Map.empty
-            Just toml -> fmap Map.fromList $ forM (HashMap.toList (tomlPairs toml)) $ \(tomlKey, anyVal) ->
-                case backward valBiMap anyVal of
-                    Right v  -> case forward keyBiMap tomlKey of
-                        Right k  -> pure (k, v)
-                        Left err -> throwError $ BiMapError err
+    input :: TomlEnv (Map k v)
+    input = asks (Prefix.lookup tableName . tomlTables) >>= \case
+        Nothing -> pure Map.empty
+        Just toml -> local (const toml) $ do
+            valKeys <- asks (HashMap.keys . tomlPairs)
+            tableKeys <- asks (fmap (:|| []) . HashMap.keys . tomlTables)
```

Comment:
Do you also need to add this?
```haskell
tableArrayKeys <- asks (HashMap.keys . tomlTableArrays)
```

Reply:
Don't think that we should support such structures.

Reply:
Why not? I can't think of a use case off the top of my head, but I'm just curious.

Reply:
By design, TOML the language discourages nested data structures. If there are compelling use-cases, we can consider adjusting the implementation.

Reply:
Ok, fair enough. Thanks.